### DISCUSSION TITLE
intended workflow when interacting with a git repository, specifically for github PRs where upstream doesn't like you force-pushing

### DISCUSSION THREAD

**Discussion by zygoloid** [jj-vcs/jj#411] [Q&A] [answered]
My `git` workflow for github PRs is generally:

* `git checkout -b branchname` from the `main` branch, make some changes, `git push -u origin branchname`
* create a matching github PR
* receive and address some review comments, `git commit -a`, `git push`
* github PR is merged, delete my branch (or usually, forget to do so)

What is the expected corresponding `jj` workflow? The closest I've found is:

* make some changes, `jj close`, `jj branch create branchname -r @-`, `jj git push`
* create a matching github PR
* receive and address some review comments, `jj close`, `jj branch set branchname -r @-`, `jj git push --branch branchname`
* github PR is merged, delete my branch (or maybe it gets autodeleted if it's deleted on github and then I pull?)

The third step in particular seems cumbersome compared to the `git ` workflow, and requires me to keep telling `jj` which branch I'm working on. Is there a better way?

**Comment by martinvonz** [comment, marked as answer]
Good question! I know that many (probably even most) GitHub projects want you to address code review comments by adding new commits on top. These projects typically integrate PRs by merging or squashing. In this project, we instead amend commits and try to keep a single commit per logical change and then integrate PRs by rebasing. That matches what e.g. the Git and Mercurial projects require (w.r.t. keeping commits clean - the Git project integrates features by merging), and also what we do internally at Google with the Fig tool. There can still be multiple commits per PR, although I only do that to work around GitHub's poor support for stacked PRs (I still expect each commit to be reviewed on its own). Incidentally, there was a [post on GitHub's blog](https://github.blog/2022-06-30-write-better-commits-build-better-projects/) today about the commit-per-logical-change workflow. It's good to see that at least a few people at GitHub care about this workflow.

Since we use the commit-per-logical-change workflow in this project, I typically don't create a branch until I'm ready to send the commit(s) for review, which I the do with `jj git push --change @-`, which creates a branch based on the Change ID (the second, magenta-colored ID in the `jj log` output).

To address review feedback, I check out the commit, make the changes and run `jj squash/amend` and then `jj git push --change @-` again (or even `jj git push` to push all branches). If you instead want to name the branch yourself, you can use e.g. `jj branch create branchname -r @-` as you're already doing (or create the branch before `jj close` and you can skip the `-r @-` bit). When you squash/amend, it will move to the rewritten commit, so just `jj git push` will work.

Once the PR has been merged, I run `jj git fetch` to propagate the branch-deletion that happened on the remote (as you speculated it might).

Anyway, if the GitHub project you're contributing to uses the squash/merge workflow (where PRs matter and commits don't matter much), I think what you're currently doing is unfortunately the best you can currently do. Well, you could drop the ` --branch branchname` argument if there are no other branches that you *don't* want to push.

Perhaps we should add a way of having a current branch like Git has. Perhaps being able to do something like `jj git push -r <commit> --to <branchname>` to basically be a shorthand for `jj branch set <branchname> -r <commit>` and `jj git push --branch <branchname>`. Yet another option might be to implement #246 to let you do something like `jj branch set <branchname> -r @-` followed by `jj git push --branch-at @-`.

**Reply by chandlerc** [reply]
FWIW, I think our project *is* actually using a single commit as a logical change.

The problem is that the GitHub pull request *code review* system means that if we amend commits in the pull request inline comments are made much harder to work with because they stay pinned to the old commit hashes. By pushing avoiding any force push to the PR branch we maximize GitHub's ability to keep inline comments visible through the course of the review.

We insist on squash merging PRs into a single logical commit (or at the end of review, assembling some small set of logical and meaningful commits in the very rare case that a PR needs to land as more than one, and then rebasing them to merge).

So what I think would be useful is just some mode / flags / command sequence that could be somewhat tailored to cope with the particular operations needed for clean PR code reviews on GitHub as that is such a widespread review system at this point.

Anyways, when working in environments like this, being able to easily use short names I create and tat are memorable / typable for me is a *huge* help as I switch back and forth. And given that I'm pushing up to Git eventually, it also helps there for wrangling GitHub's web UI etc, and matching any raw Git users' expectations that collaborate with me on the repo.

Having a current branch seems quite nice TBH. It seems to make things more familiar coming from Git w/o compromising (too much) on the model improvements. Personally, I would then have `jj git push` just push the current branch and `jj git push --all` for pushing everything.

It feels like this would provide a pretty reasonable workflow in both cases. When ammending a single commit on a branch (or maybe the latest commit for a stacked PR):
```
jj co my_pr
<do stuff>
jj amend
jj git push
```
And for cases where you need to not amend:
```
jj co my_pr
<do stuff>
jj commit
jj git push
```

WDYT?

**Reply by arxanas** [reply]
> By pushing avoiding any force push to the PR branch we maximize GitHub's ability to keep inline comments visible through the course of the review.

One alternative is to, when pushing a branch, use the commit evolution history and synthesize some artificial commits to push up. That is, if you locally amend A into B, we would synthesize C which has parent A and the content of B and push that instead. Then you wouldn't get force pushes unless you change the base of A. You should still be able to squash-merge while keeping the Github PR flow intact.

> Anyways, when working in environments like this, being able to easily use short names I create and tat are memorable / typable for me is a huge help as I switch back and forth.

There are other options too. For example, git-branchless has a fuzzy-finding interactive commit selector and `git next`/`git prev` commands for navigating the same stack.

> Having a current branch seems quite nice TBH. It seems to make things more familiar coming from Git w/o compromising (too much) on the model improvements. Personally, I would then have jj git push just push the current branch and jj git push --all for pushing everything.

I've been working on a `git submit` command for git-branchless. My preference has been to push all branches on the stack I'm working on (where a "stack" is the set of all descendant commits of the farthest ancestor draft commit). The [git-stack](https://github.com/gitext-rs/git-stack) project instead only pushes the firstmost branch in the stack, under the presumption that reviewers will only want to review one at a time. Still, neither of these push all branches.

**Reply by martinvonz** [reply]
> We insist on squash merging PRs into a single logical commit

It sounds like your project uses logical commits only after merging then, since you said you work around the limitations of GitHub's review tool by pushing fixups on top. In this project, we force-push and constantly grumble about the review tool instead :P

> So what I think would be useful is just some mode / flags / command sequence that could be somewhat tailored to cope with the particular operations needed for clean PR code reviews on GitHub as that is such a widespread review system at this point.

Yes, I agree. I've meant to do that for a long time but I've never gotten around to it. Thanks for starting this thread and providing feedback!

> Anyways, when working in environments like this, being able to easily use short names I create and tat are memorable / typable for me is a _huge_ help as I switch back and forth. And given that I'm pushing up to Git eventually, it also helps there for wrangling GitHub's web UI etc, and matching any raw Git users' expectations that collaborate with me on the repo.

Yes, makes sense.

When I switched from Git to Mercurial many years ago, I initially used bookmarks a lot. Then I gradually used them less and now I almost never use them. I admit that I sometimes wish I had bookmarked some of my longer branches, though. Making it easier to search for commits as @arxanas suggested also sounds useful. (And I totally understand that it's also a personal preference; I'm not trying to say that we shouldn't make it easy to work with branches.)

> Having a current branch seems quite nice TBH. It seems to make things more familiar coming from Git w/o compromising (too much) on the model improvements.

Yes, I think we could add that without much trouble. I have thought about doing it many times, but since I usually work with anonymous branches myself, it just hasn't happened. I suspect the model would change so each workspace (like `git worktree`) can have either a commit or a branch checked out, very similar to how Git's HEAD works. So when there's a current/active branch, it would point to the working-copy commit. I'm not sure which commands should automatically activate or deactivate the current branch. For example, it seems obvious that `jj commit` should move it to the new working-copy commit, but `jj new` probably shouldn't. This is also related to [the plan to remove the concept of open commits](#321) (it's a long thread, but see [this comment](https://github.com/martinvonz/jj/discussions/321#discussioncomment-3060363) for how to opt in to the new behavior).

I'm also not sure what the CLI for checking out a branch should be. Git has `git checkout main` for checking out the branch and `git checkout --detach main` for checking out the commit. I think we may want to reverse that so `jj co X` always resolves X as a revset instead of behaving differently if X is a branch. `jj co -b X` could activate the branch and thus require X to be a branch. That would be quite different from `git checkout -b X` (which creates a branch), but [maybe we shouldn't copy `git checkout -b X` anyway](https://stevelosh.com/blog/2013/04/git-koans/#s2-one-thing-well).

> Personally, I would then have `jj git push` just push the current branch and `jj git push --all` for pushing everything.

I agree. Defaulting to pushing less is the safer choice. I think `git` also used to push all refs by default but they have gradually reduced it so only the current branch is pushed by default (and now even only if it matches the upstream branch's name, I think). 

> It feels like this would provide a pretty reasonable workflow in both cases. When ammending a single commit on a branch (or maybe the latest commit for a stacked PR):
> 
> ```
> jj co my_pr
> <do stuff>
> jj amend
> jj git push
> ```
> 
> And for cases where you need to not amend:
> 
> ```
> jj co my_pr
> <do stuff>
> jj commit
> jj git push
> ```
> 
> WDYT?

If we make the current branch point to the working-copy commit as I suggested above, then you'll want to run `jj describe` instead of `jj commit` in the non-amend flow. With the removal of open commits, you're able allowed to push without committing, so you'd directly push the working-copy commit.

For the amend workflow, I would think that `jj amend` moves the changes from the working copy into the parent commit, but the current branch would still point to the (now-empty) working-copy commit, which is not what the user wants there. I think the obvious alternative is that the user doesn't activate the branch (so `jj co my_pr` instead of `jj co -b my_pr` or however we decide to make it look), so the branch doesn't move forward to the working-copy commit. However, then `jj git push` for pushing the current branch won't work, since there is no current branch. Maybe the option for pushing the branch pointing to the parent commit would be useful then (#246). An alternative to checking out the commit without activating the branch would be to use `jj edit`, which we could also give a `-b` option for activating a branch. Then the changes from the PR would be in the working-copy commit and the user can just make the changes and `jj git push` afterwards. I was going to say that the disadvantage is that it's then harder to see what they had changed compared to the pushed PR, but they can just use `jj diff --from my_pr@origin`. So that seems like a pretty simple workflow too.

**Reply by arxanas** [reply]
Lately, I've been thinking that it's a mistake to have branches check-out-able at all. The only special thing about a branch is that it's updated when you make a commit. I've found that I pretty much always would prefer to defer this decision to commit-time, i.e. add some `jj close --branch` flag which you pass when you want to update the branch pointing to the current commit.

**Reply by chandlerc** [reply]
TBH, while I have *many* complaints about Git, using named branches isn't one of them... I've tried git-branchless and Mercurial, and TBH its a significant regression for me. I like my branch names.

More importantly, I think a key value proposition of a VCS like JJ is the ability to collaborate with people using Git (and services like GitLab, GitHub, etc). Given that those are heavily oriented around named branches (for better or worse), I think it is valuable to make them a well integrated part of the workflow and tooling.

Anyways, getting back to the specific options for the workflow.

Not sure I fully understand the last description. Maybe some command sequences would help me? One observation is that it'd seem good to avoid pushing the working copy commit when empty / un-described / "working-copy-open". That seems likely to be quite surprising. I like a distinction more like `jj co <branch>` vs. `jj edit <branch>` over having to pick a "default" with a flag personalyl.

**Reply by chandlerc** [reply]
> Lately, I've been thinking that it's a mistake to have branches check-out-able at all. The only special thing about a branch is that it's updated when you make a commit. I've found that I pretty much always would prefer to defer this decision to commit-time, i.e. add some `jj close --branch` flag which you pass when you want to update the branch pointing to the current commit.

FWIW, I *very* often want to know what branch I am developing against. They have different baselines and expectations. I find having that state tracked and advertised to be a significant benefit. 🤷🏻 

I suspect that the branchless model just doesn't work well for me as a user. =/ Doesn't make it bad, just maybe bad for me.

**Reply by yuja** [reply]
IIUC, active branch (or bookmark in Mercurial) is considered to be attached to both working-directory and its parent commit. This wouldn't work in JJ's model where working-directory isn't special.

> Lately, I've been thinking that it's a mistake to have branches check-out-able at all. The only special thing about a branch is that it's updated when you make a commit. I've found that I pretty much always would prefer to defer this decision to commit-time, i.e. add some jj close --branch flag 

So I'm interested in this idea. The branch to be updated can be deduced from the history so long as it is linear. I could preassign branch name to working-directory commit if I liked named branch workflow.

**Reply by martinvonz** [reply]
(back from vacation now)

> Lately, I've been thinking that it's a mistake to have branches check-out-able at all. The only special thing about a branch is that it's updated when you make a commit. I've found that I pretty much always would prefer to defer this decision to commit-time, i.e. add some `jj close --branch` flag which you pass when you want to update the branch pointing to the current commit.

I've also considered mapping other VCSs' concept of current/active branch to be whatever branch(es) point to the working-copy commit. It mostly just seemed weird that there can then be multiple current branches. It also means that the user needs to remember to pass the `--branch/-b` flag every time they close/commit.

> Not sure I fully understand the last description. Maybe some command sequences would help me?

Sure, good idea. Here's how I imagine it would look in the single-commit workflow:
```
# Edit the commit pointed to by the my_pr branch in the working copy
jj edit my_pr
<address comments>
# Compare to before the fixes
jj diff --from my_pr@origin
# Push the active branch
jj git push
```

With the multi-commit workflow and the support for active branches I proposed:
```
# Start working on a new commit on top of my_pr branch, activating the branch
jj checkout -b my_pr
<address comment 1>
# Compare to before the fixes
jj diff
# Add a description and continue addressing comments in another commit on top
jj commit -m "address comment 1"
<address comment 2>
# Compare to before the fixes
jj diff --from my_pr@origin
# Add a description. Note that if this is the last commit we're going to make,
# we don't want to use `jj commit` and create a new commit.
jj describe -m "address comment 2"
# Push the active branch
jj git push
```

With the multi-commit workflow and the `jj commit --branch` @arxanas suggested (as I understood it):
```
# Start working on a new commit on top of my_pr branch, moving the branch forward to the working-copy commit
jj checkout -b my_pr
<address comment 1>
# Compare to before the fixes
jj diff
# Add a description and continue addressing comments in another commit on top
jj commit -m "address comment 1" --branch
<address comment 2>
# Compare to before the fixes
jj diff --from my_pr@origin
# Add a description. Note that if this is the last commit we're going to make,
# we don't want to use `jj commit --branch` and move the branch forward.
jj describe -m "address comment 2"
# Push the branch pointed to by the working-copy commit
jj git push
```

@arxanas and @yuja, does that match what you were thinking?

> One observation is that it'd seem good to avoid pushing the working copy commit when empty / un-described / "working-copy-open". That seems likely to be quite surprising.

We already check that commits you're about to push have a description and don't have conflicts, but good point about also checking that they're not empty.

> I like a distinction more like `jj co <branch>` vs. `jj edit <branch>` over having to pick a "default" with a flag personalyl.

The difference between those two commands is that `jj edit <commit>` lets you edit the specified commit in the working copy, while `jj co <commit>` creates a commit on top. That's the [new semantics without open commits](#321). Maybe you understood that and you're saying you'd prefer separate commands for what I called `jj checkout` and `jj checkout -b` above?

**Reply by chandlerc** [reply]
> > Not sure I fully understand the last description. Maybe some command sequences would help me?
> 
> Sure, good idea. Here's how I imagine it would look in the single-commit workflow:

Thanks, this helps me!

> ```
> # Edit the commit pointed to by the my_pr branch in the working copy
> jj edit my_pr
> <address comments>
> # Compare to before the fixes
> jj diff --from my_pr@origin
> # Push the active branch
> jj git push
> ```

With a code review system that works well w/ this flow, it seems very clean to me FWIW.

> With the multi-commit workflow and the support for active branches I proposed:
> 
> ```
> # Start working on a new commit on top of my_pr branch, activating the branch
> jj checkout -b my_pr
> <address comment 1>
> # Compare to before the fixes
> jj diff
> # Add a description and continue addressing comments in another commit on top
> jj commit -m "address comment 1"
> <address comment 2>
> # Compare to before the fixes
> jj diff --from my_pr@origin
> # Add a description. Note that if this is the last commit we're going to make,
> # we don't want to use `jj commit` and create a new commit.
> jj describe -m "address comment 2"
> # Push the active branch
> jj git push
> ```

Makes sense, thanks.

One thing that I would find useful would be to make it relatively easy to have a floating empty commit here so that users don't have to remember to not `jj commit ...` the last set of work.

If `jj git push` didn't push the working copy commit (because its empty / undescribed) but *did* push the new commits on the "current branch", I think that would just work?

> With the multi-commit workflow and the `jj commit --branch` @arxanas suggested (as I understood it):

If this indeed is as-expected, then FWIW it seems like a pretty significant overhead for fitting into the workflow demands the current review tools force us to... (at least, for me).

> > I like a distinction more like `jj co <branch>` vs. `jj edit <branch>` over having to pick a "default" with a flag personalyl.
> 
> The difference between those two commands is that `jj edit <commit>` lets you edit the specified commit in the working copy, while `jj co <commit>` creates a commit on top. That's the [new semantics without open commits](#321). Maybe you understood that and you're saying you'd prefer separate commands for what I called `jj checkout` and `jj checkout -b` above?

I think I got the key new semantics from #321. I was talking about the `-b` option above. I find that a very subtle distinction.

In the above, the fact that `jj edit` mutates an existing commit seems pretty unsurprising to me. The fact that it points the working copy at that commit first isn't completely obvious, but makes sense and seems very likely to Do What I Want.

But at that point, `jj checkout` would do something pretty surprising to me. If I wanted to make that commit the working copy and iterate on it, I would use `jj edit`. Since I didn't, the fact that I need to pass a flag is a bit surprising.

If your goal is in large part to cater to folks coming from Git, then maybe `jj switch` having the semantics you suggest for `jj checkout -b`, and `jj edit` works as you describe?

**Reply by martinvonz** [reply]
> > With the multi-commit workflow and the support for active branches I proposed:
> > ```
> > # Start working on a new commit on top of my_pr branch, activating the branch
> > jj checkout -b my_pr
> > <address comment 1>
> > # Compare to before the fixes
> > jj diff
> > # Add a description and continue addressing comments in another commit on top
> > jj commit -m "address comment 1"
> > <address comment 2>
> > # Compare to before the fixes
> > jj diff --from my_pr@origin
> > # Add a description. Note that if this is the last commit we're going to make,
> > # we don't want to use `jj commit` and create a new commit.
> > jj describe -m "address comment 2"
> > # Push the active branch
> > jj git push
> > ```
> 
> Makes sense, thanks.
> 
> One thing that I would find useful would be to make it relatively easy to have a floating empty commit here so that users don't have to remember to not `jj commit ...` the last set of work.

Yeah, I agree that that part is annoying. I think that's also related to @yuja's comment about how active branches/bookmarks in git/hg are conceptually attached to both the working copy and the parent commit (HEAD) - `git push` is an example of where it appears attached to the parent commit.

> If `jj git push` didn't push the working copy commit (because its empty / undescribed) but _did_ push the new commits on the "current branch", I think that would just work?

It feels ugly that the remote branch points to a different commit, but I can't think of an actual problem with it. A related question is what happens after you check out another commit after pushing. It will currently get abandoned and the branch will move to its parent, so that would work well with the behavior you propose.

> > With the multi-commit workflow and the `jj commit --branch` @arxanas suggested (as I understood it):
> 
> If this indeed is as-expected, then FWIW it seems like a pretty significant overhead for fitting into the workflow demands the current review tools force us to... (at least, for me).

In a way, it seems like a pretty small overhead (just a `--branch/-b` flag), but I think the problem is that you're required to pass the flag at a point in time when you don't want to have to think about whether you're on a branch or not.

I'm torn, because it's conceptually simpler if active branches don't exist, but I agree that it seems annoying to have to remember to pass the `-b` to `jj commit`.

> > > I like a distinction more like `jj co <branch>` vs. `jj edit <branch>` over having to pick a "default" with a flag personalyl.
> > 
> > 
> > The difference between those two commands is that `jj edit <commit>` lets you edit the specified commit in the working copy, while `jj co <commit>` creates a commit on top. That's the [new semantics without open commits](#321). Maybe you understood that and you're saying you'd prefer separate commands for what I called `jj checkout` and `jj checkout -b` above?
> 
> I think I got the key new semantics from #321. I was talking about the `-b` option above. I find that a very subtle distinction.
> 
> In the above, the fact that `jj edit` mutates an existing commit seems pretty unsurprising to me. The fact that it points the working copy at that commit first isn't completely obvious, but makes sense and seems very likely to Do What I Want.
> 
> But at that point, `jj checkout` would do something pretty surprising to me. If I wanted to make that commit the working copy and iterate on it, I would use `jj edit`. Since I didn't, the fact that I need to pass a flag is a bit surprising.

Is what find surprising that one would ever want to check out a commit without activating a branch? Having mostly used anonymous branches with hg for the last ~7 years, it actually took me a long time to figure out that that's what you meant (if it even was).

> If your goal is in large part to cater to folks coming from Git, then maybe `jj switch` having the semantics you suggest for `jj checkout -b`, and `jj edit` works as you describe?

You mean `jj switch` for checking activating a branch? Would it edit that branch's target directly (like `jj edit`) or would it edit a new commit on top (like `jj checkout`)? FWIW, I imagined there would be a `jj edit -b <branch name>` as well.

**Reply by chandlerc** [reply]
> > If `jj git push` didn't push the working copy commit (because its empty / undescribed) but _did_ push the new commits on the "current branch", I think that would just work?
> 
> It feels ugly that the remote branch points to a different commit, but I can't think of an actual problem with it. A related question is what happens after you check out another commit after pushing. It will currently get abandoned and the branch will move to its parent, so that would work well with the behavior you propose.

Yeah, that's roughly what I would expect.

> > > With the multi-commit workflow and the `jj commit --branch` @arxanas suggested (as I understood it):
> > 
> > 
> > If this indeed is as-expected, then FWIW it seems like a pretty significant overhead for fitting into the workflow demands the current review tools force us to... (at least, for me).
> 
> In a way, it seems like a pretty small overhead (just a `--branch/-b` flag), but I think the problem is that you're required to pass the flag at a point in time when you don't want to have to think about whether you're on a branch or not.

For me at least, I don't know how I would even incorporate this workflow.

I honestly don't think I'm capable of getting it right. Given a tool with this behavior, I think I would have to develop scripts that tracked the state somewhere and provided the flag for me. There's just no way I would get it right otherwise.

Again, this may just be a peculiarity of mine. I'm just trying to explain where I'm personally coming from, not necessarily arguing what is likely in general.

> Is what find surprising that one would ever want to check out a commit without activating a branch? Having mostly used anonymous branches with hg for the last ~7 years, it actually took me a long time to figure out that that's what you meant (if it even was).

My main comment is more that having the distinction between "check out a commit which appens to be named with a branch name, but do nothing to the branch" vs. "check out the HEAD commit of a branch and make it active" seem such different workflows that they deserve different commands and not just a flag difference.

However, I *do* find it surprising that one would ever want to check out a commit that is the HEAD of a branch without activating it.

Don't get me wrong, I *can* see the use cases. I understand how it *could* be used. I just don't find that to be a motivating workflow. Here I'll go slightly beyond just my personal preference -- almost everyone I have worked with on a team using Git for a significant time has complained about how confusing it is to check out a commit (but not the branch). And none of them have complained about checking out the branch and making it active.

I think branches and having an active branch is actually a very friendly and intuitive model. As I said earlier, this is IMO not something to "fix" from Git, but something to adopt and improve on (for example by making the commands and surrounding models better).

The only widespread complaint I have heard on teams doing development using Git (as opposed to folks working on VCSes) has been a desire to have something more like git-worktree. The branches aren't the problem there, just wanting to have separate working copies pointing at separate branches.

🤷🏻 - maybe this is unusual, all I can report is my (very limited) experience working with LLVM and similar Git upstreams.

> > If your goal is in large part to cater to folks coming from Git, then maybe `jj switch` having the semantics you suggest for `jj checkout -b`, and `jj edit` works as you describe?
> 
> You mean `jj switch` for checking activating a branch? Would it edit that branch's target directly (like `jj edit`) or would it edit a new commit on top (like `jj checkout`)? FWIW, I imagined there would be a `jj edit -b <branch name>` as well.

Based on this discussion, if you let me design it:

- `jj edit <branch name>` would make the branch active and begin editting the head commit.
- `jj switch <branch name>` would make the branch active and create a new commit for the working copy edits.
- `jj edit --no-switch <branch name>` would begin editting the head commit w/o changing branch
- Some other command to begin a new working copy commit on top of the branch head commit w/o making it active. I guess you could use `jj checkout` for that, but it seems like it creates a footgun for anyone coming from Git w/o utility.

But this reflects the fact that I find branches actively helpful for organizing independent streams of work on a repository. So I would make the workflow fairly branch-centric left to my own devices.

**Reply by yuja** [reply]
> ```
> # Start working on a new commit on top of my_pr branch, moving the branch forward to the working-copy commit
> jj checkout -b my_pr
> <address comment 1>
> # Compare to before the fixes
> jj diff
> # Add a description and continue addressing comments in another commit on top
> jj commit -m "address comment 1" --branch
> <address comment 2>
> # Compare to before the fixes
> jj diff --from my_pr@origin
> # Add a description. Note that if this is the last commit we're going to make,
> # we don't want to use `jj commit --branch` and move the branch forward.
> jj describe -m "address comment 2"
> # Push the branch pointed to by the working-copy commit
> jj git push
> ```

My idea is slightly different. There's no explicit active branch because IMHO it wouldn't work well with JJ's working-copy revision.

```
# Start working on the existing my_pr
jj checkout my_pr
# Add commit
jj commit -m "fix 1"
# Add another commit
jj commit -m "fix 2"
...
# The fixes went well, let's finalize the last commit and move the branch.
# The nearest ancestor branch 'my_pr' will be moved to this commit.
# (If the history isn't linear, or if the nearest revision has multiple branches, 
# the command would fail.)
# Maybe this could be `jj commit` && `jj branch move -r @-`
jj commit -m "fix n" --move-branch
# Push new changes to my_pr branch
jj git push
```

```
# Start new feature branch (if you prefer named branch)
jj branch set new_pr
# Add commit: branch 'new_pr' will stay there
jj commit -m "fix 1"
# Add another commit
jj commit -m "fix 2"
...
jj commit -m "fix n" --move-branch
# Push 'new_pr' branch
jj git push
```

**Reply by arxanas** [reply]
> @arxanas and @yuja, does that match what you were thinking?

I wasn't properly considering how that would interact with the working copy commit (as git-branchless doesn't have one), but I agree with @yuja's proposal above about having a directive to move the nearest ancestor branch to the imminent commit.

> It also means that the user needs to remember to pass the --branch/-b flag every time they close/commit.

That's what I would prefer in my workflow. It doesn't matter too much from a technical perspective whether we require `--branch` and don't move branches by default, or require `--no-branch` and do move branches by default. To support Git users, we could do the latter, and have lazy-branchers like me opt in via alias or configuration option.

Maybe it makes sense to have a "Git compatibility mode"? On one hand, having dramatically different behavior based on a single option might be risky/confusing, but on the other hand, it would be a lot easier to iterate on the UI without confusing Git users. In particular, I think that branches/tags don't differ enough to warrant separate concepts, and neither do checked-out branches versus non-checked-out branches (which also lead to a host of complexity like symbolic references).

Aside: despite having gone to the trouble of inventing symbolic references, Git actually doesn't work too well with them. For example, I tried to make a symbolic reference `master` which pointed to `origin/master`, but this terribly confuses Git, and it starts updating the local/remote branch in places where it shouldn't. Git is pretty much only equipped to handle the case of `HEAD` being a symbolic reference, and that reference has to have no more than one layer of symbolic indirection. Here's another confusing case where Git inappropriately resolves a symbolic reference: https://github.com/arxanas/git-branchless/discussions/149#discussioncomment-1526830

Aside: in https://github.com/martinvonz/jj/discussions/321#discussioncomment-2832646, I give a theoretical justification for inferring branches: "...history edges let us infer logical branches, without having to name those branches explicitly. Instead, the branches are determined by the operations you do on the commits." The ability to name branches is still useful, but naming things doesn't work well for conflict resolution in a distributed context. This is why I suggested moving the branches only when you make a new commit: because it matches the model of determining branches via the operations you do on the commits, rather than as entirely separate entities with their own rules.

**Reply by martinvonz** [reply]
From @chandlerc:

> Don't get me wrong, I _can_ see the use cases. I understand how it _could_ be used. I just don't find that to be a motivating workflow. Here I'll go slightly beyond just my personal preference -- almost everyone I have worked with on a team using Git for a significant time has complained about how confusing it is to check out a commit (but not the branch). And none of them have complained about checking out the branch and making it active.

Right, but I believe that the reason they think so is because of how scary Git makes it. There's ~10 lines of instructions for how to escape for from the so-called "detached HEAD" state (a scary name as well). There's of course good reason for that, because it's easy to lose commits that way if you're not experienced enough to use the reflog. Jujutsu (and git-branchless) doesn't have that problem, because anonymous heads are tracked and show up in log output. Despite Git's loud warnings, there are actually people who develop mostly using detached HEAD (I know someone on the Git team at Google who does). I was quite surprised when I learned that six months ago or so. Mercurial also uses anonymous branching in its [tutorial](https://www.mercurial-scm.org/wiki/TutorialFirstChange).

That said, making the transition as easy as possible for Git users is a very important goal. I'm not saying we should de-prioritize named-branch workflows. I'm just saying that anonymous branching is probably something you'll get used to relatively quickly. As a data point, I think it's actually very few Fig/Mercurial users at Google who use branches/bookmarks.

> The only widespread complaint I have heard on teams doing development using Git (as opposed to folks working on VCSes) has been a desire to have something more like git-worktree. The branches aren't the problem there, just wanting to have separate working copies pointing at separate branches.

But git-worktree does exist (as you clearly know), so what are they complaining about? Just that they were not aware of the feature? (Pointing them to separate branches is the only allowed way of using them, IIUC - you're not allowed to check out the same branch in two worktrees.)

From @arxanas:

> That's what I would prefer in my workflow. It doesn't matter too much from a technical perspective whether we require `--branch` and don't move branches by default, or require `--no-branch` and do move branches by default. To support Git users, we could do the latter, and have lazy-branchers like me opt in via alias or configuration option.

I agree. Defaulting to moving forward any branch(es?) on the working-copy commit sounds good. For people using a branchless workflow, that would still work well if we simply make it allowed to run `jj commit` without `--no-branch` (@yuja's `--[no-]move-branch` is probably a better name) even if there is no branch. I think I like that proposal, so thank you.

So, that would mean this sequence of commands for the named-branch workflow:

```
# Check out a new commit on top of my_pr and move my_pr to it
jj switch my_pr
# Add commit, moving the branch forward
jj commit -m "fix 1"
# Add another commit, moving the branch forward again
jj commit -m "fix 2"
# Push new changes to my_pr branch, but since `@` is empty, pushes `@-` instead
jj git push
```

To be clear, this is the no-active-branch solution @yuja and @arxanas suggested. A branch that points to the working-copy commit simply behaves very similar to an active branch in git/hg. This solution means `jj edit <branch>` would always make it "active" (by virtue of pointing to the working-copy commit), and there would be no `jj edit --no-branch`, which seems like a good simplification to me.

Maybe we still want a way of creating a new branch and switching to it with a single command, like `git checkout -b <new-branch> <start-point>`, but that's a detail we can decide on later.

**Reply by yuja** [reply]
> ```
> # Check out a new commit on top of my_pr and move my_pr to it
> jj switch my_pr
> # Add commit, moving the branch forward
> jj commit -m "fix 1"
> # Add another commit, moving the branch forward again
> jj commit -m "fix 2"
> # Push new changes to my_pr branch, but since `@` is empty, pushes `@-` instead
> jj git push
> ```

Pushing `@-` instead of `@` seems fine, but after that, the history would look like:
```
@ (no description set) my_pr
o "fix 2" my_pr@origin
```

Doesn't it lead to future branch conflicts?

1. PR merged and the `my_pr` branch gets deleted by the remote.
2. `jj git fetch` detects deletion of `my_pr`, but our `my_pr` points to `@`, not to `@-`.

We could special case a branch pointing to an empty head, but I'm afraid that would cause another confusing behavior.

**Reply by martinvonz** [reply]
> Doesn't it lead to future branch conflicts?

Good catch! Perhaps we could move the local branch to `@-` as an early step in `jj git push`, so the local branch and remote branch would point to the same commit. That would be like detaching the branch (in Git speak),  but maybe that's fine as long as we tell the user about it so it's not too surprising.
